### PR TITLE
[MPS] Handle in_features=0 in linear forward and backward

### LINKED ISSUE #190050: [MPS] linear with in_features=0 aborts the process; backward also fails
### 🐛 Describe the bug

`F.linear` with `in_features = 0` hard-crashes the whole process with SIGABRT
on macOS 15+. CPU (and CUDA) return the broadcast bias, or zeros without bias.
The MPS forward only early-returns when the *output* is empty, which it is not
for `K = 0, N > 0`, and then builds an MPSNDArray over a zero-length dimension,
which trips an uncatchable MetalPerformanceShaders assertion.

```python
import torch, torch.nn.functional as F

x = torch.randn(3, 4, 0)
w = torch.randn(5, 0)
b = torch.randn(5)
print(F.linear(x, w, b).shape)  # CPU: (3, 4, 5), equals broadcast bias

out = F.linear(x.to("mps"), w.to("mps"), b.to("mps"))  # SIGABRT
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable:

```text
torch.Size([3, 4, 5])
MPSNDArray.mm:759: failed assertion `[MPSNDArray initWithDevice:descriptor:isTextureBacked:] Error: device may not be nil'
[process aborts, rc=-6]
```

`nn.Linear(0, 5)` crashes the same way; bias is not required. `torch.matmul`
and `torch.addmm` with `K = 0` on MPS are fine — only linear's macOS-15+
no-graph fast path is unguarded (the pre-macOS-15 MPSGraph path fails loudly
with a catchable "Placeholder tensor is empty!" instead).

Backward is independently broken: the input-gradient path checks
`grad_output.numel()` rather than the (empty) grad-input, and the weight path
reshapes and runs a graph for the empty `(N, 0)` weight gradient, so direct
`aten.linear_backward` calls for input, weight, and bias masks all raise.
CPU returns valid empty gradients and `grad_bias = grad_output.sum(...)`.

The forward guard is at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/Linear.mm#L119-L121
and the backward guards at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/Linear.mm#L238-L240
and
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/Linear.mm#L311-L315
check only `grad_output`.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @ezyang @gchanan @kadeng @msaroufim @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
Forward only early-returned for an empty *output*, which a `K=0, N>0` linear does not have, and then built an MPSNDArray over the zero-length dimension, hitting an uncatchable MetalPerformanceShaders assertion that aborts the whole process on macOS 15+ (the pre-15 MPSGraph path failed loudly instead). Return the broadcast bias (or zeros without bias) instead, matching CPU.

Backward guarded only on `grad_output.numel()`: the input-gradient path ran a graph toward an empty grad-input, and the weight path hit an ambiguous `reshape({-1, 0})` on the empty input before producing the empty `(N, 0)` weight gradient and the still-nonzero bias gradient. Handle empty input/grad_output before the reshapes; `grad_bias` remains the sum of grad_output over the leading dims.

The new `test_linear_zero_in_features` runs forward and backward with and without bias and compares outputs and all gradients against CPU. The repro from the issue (child process) now exits cleanly where it previously died with SIGABRT. All 37 linear tests pass.

Fixes #190050

*This PR was authored with assistance from Claude.*

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -1532,6 +1532,27 @@ def helper(bias_shape):
...
+                m_mps = copy.deepcopy(m_cpu).to("mps")
+                out_mps = m_mps(x_mps)
+                out_mps.backward(torch.ones_like(out_mps))
+
+                self.assertEqual(out_mps.cpu(), out_cpu)
+                self.assertEqual(x_mps.grad.cpu(), x_cpu.grad)
+                self.assertEqual(m_mps.weight.grad.cpu(), m_cpu.weight.grad)
+                if has_bias:
+                    self.assertEqual(m_mps.bias.grad.cpu(), m_cpu.bias.grad)
+
```
Comment (malfet) [inline test/test_mps.py]: Can you add this one to OpInfo for linear (i.e. it's not MPS specific, should work the same way on CPU or CUDA, right?

**Review by malfet — APPROVED** [review]
LGTM, please note how I've refactored a change to use OpInfo instead of just adding an MPS specific test, as fix should apply uniformly to all devices

File: test/test_mps.py (hunk 1 above)
Comment (bogatyy) [reply test/test_mps.py]: Oh, I did not think of that, thank you.

Comment (malfet) [reply test/test_mps.py]: Well, the downside is that one is likely to discover similar problems in other backend, one such problem are bugs in forward AD, let me address those as well
